### PR TITLE
Make ConnectionPool cache extra connections for a while.

### PR DESCRIPTION
**Pull request by AlCutter**
Try to handle connection churn (e.g. from etcd deletes) a bit better.

### CODE REVIEW

File: cpp/net/connection_pool.cc [hunk 1]
```diff
@@ -240,9 +256,9 @@ unique_ptr<ConnectionPool::Connection> ConnectionPool::Get(const URL& url) {
   }
 
   VLOG(1) << "cached evhtp_connection for " << key.first << ":" << key.second;
-  unique_ptr<ConnectionPool::Connection> retval(move(it->second.back()));
+  unique_ptr<ConnectionPool::Connection> retval(
+      move(it->second.back().second));
```
Comment (phad) [inline cpp/net/connection_pool.cc]: What if there's no second.back()

File: cpp/net/connection_pool.cc [hunk 2]
```diff
@@ -279,14 +298,22 @@ void ConnectionPool::Put(unique_ptr<ConnectionPool::Connection>&& conn) {
 void ConnectionPool::Cleanup() {
   unique_lock<mutex> lock(lock_);
   cleanup_scheduled_ = false;
+  const system_clock::time_point cutoff(
+      system_clock::now() -
+      seconds(FLAGS_connection_pool_max_unused_age_seconds));
 
-  // std::map<HostPortPair, std::deque<unique_ptr<Connection>>> conns_;
+  // std::map<HostPortPair, std::deque<TimestampedConnection>> conns_;
```
Comment (phad) [inline cpp/net/connection_pool.cc]: Remove this commented out stuff

File: cpp/net/connection_pool.cc [hunk 3]
```diff
@@ -266,7 +282,10 @@ void ConnectionPool::Put(unique_ptr<ConnectionPool::Connection>&& conn) {
   auto& entry(conns_[key]);
 
   CHECK_GE(FLAGS_url_fetcher_max_conn_per_host_port, 0);
-  entry.emplace_back(move(conn));
+  entry.emplace_back(make_pair(system_clock::now(), move(conn)));
```
Comment (phad) [inline cpp/net/connection_pool.cc]: If you were to inject a system_clock this would be more easily tested.

File: cpp/net/connection_pool.cc (hunk 1 above)
Comment (AlCutter) [inline cpp/net/connection_pool.cc]: There has to be at this point, see L244

File: cpp/net/connection_pool.cc (hunk 2 above)
Comment (AlCutter) [inline cpp/net/connection_pool.cc]: I think that was put there as a reference for what the `auto` in the for loop below would be.

File: cpp/net/connection_pool.cc (hunk 3 above)
Comment (AlCutter) [inline cpp/net/connection_pool.cc]: Ack.

File: cpp/net/connection_pool.cc (hunk 2 above)
Comment (phad) [inline cpp/net/connection_pool.cc]: Remove the conns_; bit then so it looks less like commented-out code.

Comment (AlCutter) [inline cpp/net/connection_pool.cc]: oke doke

**Pull request merged**